### PR TITLE
Set GitHub usernames via oauth to GitHub.

### PR DESCRIPTION
**Pull request by benjaminp**
Replace the free-form GitHub username form with a button that runs an OAuth flow with GitHub to determine a users's GitHub username. Both administrators and users can clear a GitHub username from the database with the "Clear" button.

There is now a configuration key for the GitHub APP client id to authenticate to the user. A plugin must be provided to give the GitHub app's secret.

Some deployments may need to talk to the GitHub API through a proxy. Configuration variables for proxy host and port and added to support this. Grouper must also depend on pycurl now, too, because only the curl tornado client supports proxies.

### CODE REVIEW

File: grouper/fe/routes.py [hunk 1]
```diff
@@ -81,6 +81,7 @@
     (r"/audits", AuditsView),
     (r"/audits/(?P<audit_id>[0-9]+)/complete", AuditsComplete),
     (r"/audits/create", AuditsCreate),
+    (r"/github/link/(?P<user_id>[0-9]+)", GitHubLinkView),
```
[the remark this answers is not in the document]
Comment (benjaminp) [reply grouper/fe/routes.py]: It has to be this way because Grouper matches the authorization callback against a prefix.

File: tests/github_test.py [hunk 2]
```diff
@@ -18,42 +26,119 @@
...
+
+
+class SecretPlugin(object):
+    def get_github_app_client_secret(self):
+        return b"client-secret"
+
+
 @pytest.mark.gen_test
-def test_github(session, users, http_client, base_url):  # noqa: F811
+def test_github(session, users, http_client, base_url, monkeypatch):  # noqa: F811
```
[the remark this answers is not in the document]
Comment (benjaminp) [reply tests/github_test.py]: I know, but I was hoping to start a trend of using a nicer tool.

File: grouper/fe/handlers/github.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        user = User.get(self.session, user_id)
+        if not user:
+            self.notfound()
+            return
+        if self.current_user.id != user.id:
+            self.forbidden()
+            return
+
+        # Turn the temporary oauth code into an access token.
+        github_client = _get_github_client()
```
[the remark this answers is not in the document]
Comment (benjaminp) [reply grouper/fe/handlers/github.py]: You're thinking something like this?
```
class GitHub:
    def login(self):...
    def get_username(self, token):...
```
I want the fake testing GitHub interface to be raw HTTP not something like above. I don't think Merou needs a generic GitHub client.

Normally, I would put what's currently in `get()` into a separate function and pass the HTTP client to it, so patching wasn't required. However, in this case, I want everything from the tornado web-frameworky code to the GitHub HTTP client integration tested.

In summary, this code was written in a deliberately inflexible way to minimize the places it could bend and break.

File: grouper/fe/routes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cbguder) [reply grouper/fe/routes.py]: Ah, that's interesting. TIL.

File: grouper/fe/handlers/github.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cbguder) [reply grouper/fe/handlers/github.py]: Yep, that's roughly what I had in mind. I think we can keep the tests the same and still do this refactor? It would make the code more readable and maintainable.

[the remark this answers is not in the document]
Comment (benjaminp) [reply grouper/fe/handlers/github.py]: Sure, see latest commit.

File: grouper/fe/routes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benjaminp) [reply grouper/fe/routes.py]: I suppose I could configure GitHub's prefix to be simply `/` but more specific seems better.

File: grouper/fe/handlers/github.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cbguder) [reply grouper/fe/handlers/github.py]: Awesome, thanks Benjamin!

File: config/dev.yaml [hunk 4]
```diff
@@ -29,6 +29,22 @@ common:
...
 
+    # "client id" of a GitHub app used to find GitHub usernames. A plugin must
+    # provide the client secret.
+    #
+    # Type: str
+    github_app_client_id: null
+
+    # Host for the proxy that external HTTP requests are made over.
+    #
+    # Type: int
```
Comment (wrodriguez-dbx) [inline config/dev.yaml]: int -> str?

Comment (benjaminp) [reply config/dev.yaml]: Yep, good catch!

**Comment by rra** [comment]
I left this for too long, so there were various merge changes required. I've put up https://github.com/dropbox/merou/pull/595 which includes the same base commit plus those additional fixes.
